File: LeetCode/1_Two_Sum.cpp
// Problem Link: https://leetcode.com/problems/two-sum/
// Intuition: We are given an array of integers and a target value. Our goal is to find two numbers in the array that add up to the target.
// To solve this, we use a hashmap to store the values we have seen so far along with their indices. As we iterate through the array,
// for each element, we calculate the remaining value needed to reach the target. If the remaining value exists in the hashmap,
// we have found our pair, and we return their indices. If no pair is found, we continue the iteration.

// Time Complexity: O(n), where n is the number of elements in the array.
// Space Complexity: O(n) due to the use of a hashmap.

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        int n = nums.size();
        unordered_map<int, int> index_map; // Create a hashmap to store values and their indices.

        for (int i = 0; i < n; i++) {
            int remaining = target - nums[i]; // Calculate the remaining value needed to reach the target.

            // Check if the remaining value exists in the hashmap.
            if (index_map.find(remaining) != index_map.end()) {
                // If found, return the indices of the two numbers that add up to the target.
                return {index_map[remaining], i};
            } else {
                // Otherwise, store the current number and its index in the hashmap.
                index_map[nums[i]] = i;
            }
        }

        // Return an invalid pair if no solution is found.
        return {-1, -1};
    }
};

File: LeetCode/1_two_sum.py
'''
Question URL: https://leetcode.com/problems/two-sum/
Approach:
1. Start by copy and sorting.
2. Then, we find two numbers with two pointers.
3. Finally, return two index numbers from origin list 'nums'.
Time Complexity: O(NlogN) (maybe)
Space Complexity: O(N) (maybe)
'''

class Solution:
    def find_nth_occurrence(self, lst, number, n):
        count = 0
        for i, element in enumerate(lst):
            if element == number:
                count += 1
                if count == n:
                    return i
        return None
    
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        nums_copy = nums.copy()
        nums_copy.sort()

        left = 0
        right = len(nums_copy) - 1
        while left < right:
            current_sum = nums_copy[left] + nums_copy[right]
            if current_sum == target:
                if nums_copy[left] == nums_copy[right]:
                    return [self.find_nth_occurrence(nums, nums_copy[left], 1), 
                            self.find_nth_occurrence(nums, nums_copy[left], 2)]
                else:
                    return [nums.index(nums_copy[left]), nums.index(nums_copy[right])]
            elif current_sum < target:
                left += 1
            else:
                right -= 1

        return None
